File: 0036-valid-sudoku/0036-valid-sudoku.cpp
class Solution {
public:
    bool isValidSudoku(vector<vector<char>>& board) {
        vector<vector<int>> rows(9, vector<int>(9, 0));
        vector<vector<int>> columns(9, vector<int>(9, 0));
        vector<vector<int>> grid(9, vector<int>(9, 0));

        for (int i = 0; i < board.size(); i++) {
            for (int j = 0; j < board[0].size(); j++) {
                if (board[i][j] != '.') {
                    int num = board[i][j] - '0';
                    int k = i / 3 * 3 + j / 3;
                    if (rows[i][num - 1]++|| columns[j][num - 1]++ || grid[k][num - 1]++)
                    {
                        return false;
                    }
                }
            }
        }
        return true;
    }
};

File: 0036-valid-sudoku/0036-valid-sudoku.rs
impl Solution {
    pub fn is_valid_sudoku(board: Vec<Vec<char>>) -> bool {
        let mut rows: Vec<Vec<i32>> = vec![vec![0; 9]; 9];
        let mut columns: Vec<Vec<i32>> = vec![vec![0; 9]; 9];
        let mut grid: Vec<Vec<i32>> = vec![vec![0; 9]; 9];

        for i in 0..board.len() {
            for j in 0..board[0].len() {
                if board[i][j] != '.' {
                    let num = (board[i][j] as u8 - b'0') as usize;
                    let k = i / 3 * 3 + j / 3;

                if rows[i][num - 1] > 0 || columns[j][num - 1] > 0 || grid[k][num - 1] > 0 {
                    return false;
                }
                
                rows[i][num - 1] += 1;
                columns[j][num - 1] += 1;
                grid[k][num - 1] += 1;
                }
            }
        }
        true
    }
}